File: rxbinding-design/src/main/java/com/jakewharton/rxbinding/support/design/widget/RxSnackbar.java
package com.jakewharton.rxbinding.support.design.widget;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;

import rx.Observable;

import static com.jakewharton.rxbinding.internal.Preconditions.checkNotNull;

/**
 * Static factory methods for creating {@linkplain Observable observables} for {@link Snackbar}.
 */
public final class RxSnackbar {
  /**
   * Create an observable which emits the dismiss events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<Integer> dismisses(@NonNull Snackbar view) {
    checkNotNull(view, "view == null");
    return Observable.create(new SnackbarDismissesOnSubscribe(view));
  }
  /**
   * Create an observable which emits the action clicked events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<Integer> actionClicked(@NonNull Snackbar view, int resId) {
    checkNotNull(view, "view == null");
    return Observable.create(new SnackbarActionOnSubscribe(view, resId));
  }
  /**
   * Create an observable which emits the action clicked events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<Integer> actionClicked(@NonNull Snackbar view, CharSequence text) {
    checkNotNull(view, "view == null");
    return Observable.create(new SnackbarActionOnSubscribe(view, text));
  }

  private RxSnackbar() {
    throw new AssertionError("No instances.");
  }
}

File: rxbinding-design-kotlin/src/main/kotlin/com/jakewharton/rxbinding/support/design/widget/RxSnackbar.kt
package com.jakewharton.rxbinding.support.design.widget

import android.support.design.widget.Snackbar
import rx.Observable

/**
 * Create an observable which emits the dismiss events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.dismisses(): Observable<Int> = RxSnackbar.dismisses(this)

/**
 * Create an observable which emits the action clicked events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.actionClicked(resId: Int): Observable<Int> = RxSnackbar.actionClicked(this, resId)

/**
 * Create an observable which emits the action clicked events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.actionClicked(text: CharSequence): Observable<Int> = RxSnackbar.actionClicked(this, text)
